File: Hard/Longest repeating and non-overlapping substring/longest-repeating-and-nonoverlapping-substring.java
//{ Driver Code Starts
//Initial Template for Java

import java.io.*;
import java.util.*;

class GFG {
    public static void main(String args[]) throws IOException {
        BufferedReader read =
            new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(read.readLine());
        while (t-- > 0) {
            int N = Integer.parseInt(read.readLine());
            String S = read.readLine();

            Solution ob = new Solution();
            System.out.println(ob.longestSubstring(S, N));
        }
    }
}

// } Driver Code Ends


//User function Template for Java

class Solution {
    static String longestSubstring(String s, int n) {
        // code here
        int window = 1;
        StringBuilder str = new StringBuilder();
        String ans = "";
        
        for(int i = 0; i<s.length(); i++) {
            while(str.length() != window) str.append(s.charAt(i));
            
            if(s.indexOf(str.toString(), i+1) != -1) {
                window++;
                ans = str.toString();
            }
            else {
                str.deleteCharAt(0);
            }
        }
        
        return ans.length() == 0 ? "-1" : ans;
    }
};

File: Hard/Longest repeating and non-overlapping substring/longest-repeating-and-nonoverlapping-substring.py
#User function Template for python3

class Solution:
    def longestSubstring(self, s , n):
        # code here 
        window = 1
        str = ""
        ans = ""
    
        for i in range(n):
            while len(str) != window:
                str += s[i]
    
            if s.find(str, i + 1) != -1:
                window += 1
                ans = str
            else:
                str = str[1:]
    
        if len(ans) == 0:
            return "-1"
        return ans


#{ 
 # Driver Code Starts
#Initial Template for Python 3

if __name__ == '__main__': 
    t = int (input ())
    for _ in range (t):
        N=int(input())
        S=input()
        
        ob = Solution()
        print(ob.longestSubstring(S , N))
# } Driver Code Ends